File: cses/09_geometry/1741_area-of-rectangles/solution.py
import sys


def main():
    data = sys.stdin.buffer.read().split()
    nums = list(map(int, data))
    pos = 0
    n = nums[pos]; pos += 1

    X1 = [0] * n
    Y1 = [0] * n
    X2 = [0] * n
    Y2 = [0] * n
    ys_set = set()
    for i in range(n):
        x1 = nums[pos]; y1 = nums[pos + 1]; x2 = nums[pos + 2]; y2 = nums[pos + 3]
        pos += 4
        X1[i] = x1; Y1[i] = y1; X2[i] = x2; Y2[i] = y2
        ys_set.add(y1)
        ys_set.add(y2)

    # Coordinate compression trục y.
    ys = sorted(ys_set)
    yidx = {v: i for i, v in enumerate(ys)}
    M = len(ys) - 1  # số elementary y-interval (>=1 vì luôn có y1<y2)

    # Kích thước segment tree (luỹ thừa 2 >= M).
    sz = 1
    while sz < M:
        sz <<= 1

    # span[node] = tổng chiều dài vật lý theo y của các elementary interval trong node.
    span = [0] * (2 * sz)
    for i in range(M):
        span[sz + i] = ys[i + 1] - ys[i]
    for i in range(sz - 1, 0, -1):
        span[i] = span[2 * i] + span[2 * i + 1]

    # cnt[node]   = số hình chữ nhật active phủ trọn đoạn node (lazy, không push-down).
    # cover[node] = độ dài phần bị phủ trong đoạn node (tính theo cnt tại node và hậu duệ).
    cnt = [0] * (2 * sz)
    cover = [0] * (2 * sz)

    # Dựng các sự kiện dọc: tại x1 cộng +1 trên [y1,y2), tại x2 cộng -1 trên [y1,y2).
    events = []
    ap = events.append
    for i in range(n):
        li = yidx[Y1[i]]
        ri = yidx[Y2[i]]
        ap((X1[i], 1, li, ri))
        ap((X2[i], -1, li, ri))
    events.sort()

    area = 0
    ne = len(events)
    prev_x = events[0][0]
    j = 0
    while j < ne:
        x = events[j][0]
        # Dải [prev_x, x] được phủ với chiều cao = độ dài union hiện tại cover[1].
        area += (x - prev_x) * cover[1]
        # Áp dụng mọi sự kiện có cùng hoành độ x.
        while j < ne and events[j][0] == x:
            _, val, l, r = events[j]
            j += 1
            # Range update +val/-val trên các elementary interval [l, r), duy trì cover
            # bằng công thức cục bộ:
            #   cover = span            nếu cnt > 0
            #   cover = 0               nếu là lá và cnt == 0
            #   cover = con0 + con1     trong các trường hợp còn lại
            l += sz
            r += sz
            ll = l
            rr = r
            while l < r:
                if l & 1:
                    c = cnt[l] + val
                    cnt[l] = c
                    if c > 0:
                        cover[l] = span[l]
                    elif l >= sz:
                        cover[l] = 0
                    else:
                        cover[l] = cover[2 * l] + cover[2 * l + 1]
                    l += 1
                if r & 1:
                    r -= 1
                    c = cnt[r] + val
                    cnt[r] = c
                    if c > 0:
                        cover[r] = span[r]
                    elif r >= sz:
                        cover[r] = 0
                    else:
                        cover[r] = cover[2 * r] + cover[2 * r + 1]
                l >>= 1
                r >>= 1
            # Tính lại cover cho toàn bộ tổ tiên của hai biên lên tới gốc.
            k = ll >> 1
            while k >= 1:
                if cnt[k] > 0:
                    cover[k] = span[k]
                else:
                    cover[k] = cover[2 * k] + cover[2 * k + 1]
                k >>= 1
            k = (rr - 1) >> 1
            while k >= 1:
                if cnt[k] > 0:
                    cover[k] = span[k]
                else:
                    cover[k] = cover[2 * k] + cover[2 * k + 1]
                k >>= 1
        prev_x = x

    sys.stdout.write(str(area) + "\n")


main()

File: cses/09_geometry/1741_area-of-rectangles/solution.cpp
#include <algorithm>
#include <cstdint>
#include <iostream>
#include <vector>

// Sự kiện dọc của đường quét: tại hoành độ x, thêm/bớt (delta = +1/-1) độ phủ
// trên đoạn y [y_low, y_high).
struct Event {
    int x;
    int y_low;
    int y_high;
    int delta;
};

// Segment tree trên các toạ độ y đã coordinate-compress, duy trì tổng độ dài y
// đang bị phủ bởi các hình chữ nhật active.
class CoverageTree {
public:
    explicit CoverageTree(const std::vector<int>& coordinates)
        : ys_(coordinates), interval_count_(static_cast<int>(coordinates.size()) - 1),
          count_(static_cast<std::size_t>(4 * interval_count_), 0),
          covered_(static_cast<std::size_t>(4 * interval_count_), 0) {}

    // Range update +delta/-delta trên các elementary interval [query_left, query_right].
    void update(int query_left, int query_right, int delta) {
        update(1, 0, interval_count_ - 1, query_left, query_right, delta);
    }

    // Tổng độ dài y đang phủ toàn cục = cover tại gốc (node 1).
    [[nodiscard]] std::int64_t covered_length() const {
        return covered_[1];
    }

private:
    const std::vector<int>& ys_;
    int interval_count_;
    std::vector<int> count_;       // cnt[node]: số hình chữ nhật phủ trọn đoạn node (lazy, không push-down)
    std::vector<std::int64_t> covered_;  // cover[node]: độ dài phủ trong đoạn node

    // Công thức truy hồi cục bộ cho cover.
    void pull(int node, int left, int right) {
        if (count_[static_cast<std::size_t>(node)] > 0) {
            // cnt > 0: cả đoạn bị phủ trọn -> cover = span vật lý theo y.
            covered_[static_cast<std::size_t>(node)] =
                static_cast<std::int64_t>(ys_[static_cast<std::size_t>(right + 1)]) -
                ys_[static_cast<std::size_t>(left)];
        } else if (left == right) {
            // Lá và cnt == 0: không phủ.
            covered_[static_cast<std::size_t>(node)] = 0;
        } else {
            // Còn lại: gộp từ hai con.
            covered_[static_cast<std::size_t>(node)] =
                covered_[static_cast<std::size_t>(2 * node)] +
                covered_[static_cast<std::size_t>(2 * node + 1)];
        }
    }

    void update(int node, int left, int right, int query_left, int query_right, int delta) {
        if (query_left <= left && right <= query_right) {
            // Đoạn node nằm trọn trong khoảng cập nhật: chỉ đổi cnt rồi tính lại cover.
            count_[static_cast<std::size_t>(node)] += delta;
            pull(node, left, right);
            return;
        }

        const int middle = left + (right - left) / 2;
        if (query_left <= middle) {
            update(2 * node, left, middle, query_left, query_right, delta);
        }
        if (query_right > middle) {
            update(2 * node + 1, middle + 1, right, query_left, query_right, delta);
        }
        pull(node, left, right);
    }
};

int main() {
    std::ios::sync_with_stdio(false);
    std::cin.tie(nullptr);

    int rectangle_count = 0;
    std::cin >> rectangle_count;

    std::vector<Event> events;
    events.reserve(static_cast<std::size_t>(2 * rectangle_count));
    std::vector<int> ys;
    ys.reserve(static_cast<std::size_t>(2 * rectangle_count));

    // Mỗi hình chữ nhật sinh hai sự kiện dọc: +1 tại cạnh trái x1, -1 tại cạnh phải x2.
    for (int i = 0; i < rectangle_count; ++i) {
        int x1 = 0;
        int y1 = 0;
        int x2 = 0;
        int y2 = 0;
        std::cin >> x1 >> y1 >> x2 >> y2;
        events.push_back({x1, y1, y2, 1});
        events.push_back({x2, y1, y2, -1});
        ys.push_back(y1);
        ys.push_back(y2);
    }

    // Coordinate compression trục y và sắp các sự kiện theo x tăng dần.
    std::sort(ys.begin(), ys.end());
    ys.erase(std::unique(ys.begin(), ys.end()), ys.end());
    std::sort(events.begin(), events.end(), [](const Event& first, const Event& second) {
        return first.x < second.x;
    });

    CoverageTree tree(ys);
    std::int64_t area = 0;
    int previous_x = events.front().x;
    std::size_t event_index = 0;

    // Quét từng nhóm sự kiện cùng hoành độ.
    while (event_index < events.size()) {
        const int x = events[event_index].x;
        // Cộng diện tích dải [previous_x, x] TRƯỚC khi áp dụng sự kiện tại x hiện tại.
        area += static_cast<std::int64_t>(x - previous_x) * tree.covered_length();

        // Áp dụng mọi sự kiện có cùng hoành độ x.
        while (event_index < events.size() && events[event_index].x == x) {
            const Event& event = events[event_index];
            // Ánh xạ [y_low, y_high) sang chỉ số elementary interval [left, right].
            const int left = static_cast<int>(
                std::lower_bound(ys.begin(), ys.end(), event.y_low) - ys.begin());
            const int right = static_cast<int>(
                std::lower_bound(ys.begin(), ys.end(), event.y_high) - ys.begin()) - 1;
            tree.update(left, right, event.delta);
            ++event_index;
        }
        previous_x = x;
    }

    std::cout << area << '\n';
    return 0;
}
